wire-ios-data-model/Source/Proteus/ProteusProvider.swift

```diff
@@ -0,0 +1,70 @@
+//
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireUtilities
+
+public typealias KeyStorePerformBlock<T> = ((UserClientKeysStore) throws -> T)
+public typealias ProteusServicePerformBlock<T> = ((ProteusServiceInterface) throws -> T)
+
+public protocol ProteusProviding {
+
+    func perform<T>(
+        withProteusService proteusServiceBlock: ProteusServicePerformBlock<T>,
+        withKeyStore keyStoreBlock: KeyStorePerformBlock<T>
+    ) rethrows -> T
+
+}
+
+public class ProteusProvider: ProteusProviding {
+
+    private let context: NSManagedObjectContext
+    private let proteusViaCoreCrypto: Bool
+
+    public init(
+        context: NSManagedObjectContext,
+        proteusViaCoreCrypto: Bool = DeveloperFlag.proteusViaCoreCrypto.isOn
```

Comment:
what happens if the developerFlag is toggle? Is ProteusProvider recreated somehow?

Reply:
no, but I don't know if it's necessary